Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+      foreach (var constraint in unnormalizedSubtypeConstraints) {
+        if (0 <= direction && constraint.Super.Normalize() == tproxy) {
+          yield return constraint.Sub;
+        }
+        if (direction <= 0 && constraint.Sub.Normalize() == tproxy) {
+          yield return constraint.Super;
+        }
+      }
+    }
+
+    // ---------------------------------------- Guarded constraints ----------------------------------------
+
+    void AddGuardedConstraint(Func<bool> predicate) {
+      guardedConstraints.Add(predicate);
+    }
+
+    bool ApplyGuardedConstraints() {
+      if (guardedConstraints.Count == 0) {
+        // common special case
+        return false;
+      }
+      var constraints = guardedConstraints;
+      guardedConstraints = new();
+      var anythingChanged = false;
+      foreach (var constraint in constraints) {
+        if (constraint()) {
+          anythingChanged = true;
+        } else {
+          guardedConstraints.Add(constraint);
+        }
+      }
+      return anythingChanged;
+    }
+
+    // ---------------------------------------- Advice ----------------------------------------
+
+    void AddDefaultAdvice(PreType preType, Advice.Target advice) {
+      defaultAdvice.Add(new Advice(preType, advice));
+    }
+
+    bool TryApplyDefaultAdvice() {
+      bool anythingChanged = false;
+      foreach (var advice in defaultAdvice) {
+        anythingChanged |= advice.Apply(PreTypeResolver);
+      }
+      return anythingChanged;
+    }
+
+    // ---------------------------------------- Post-inference confirmations ----------------------------------------
+
+    void AddConfirmation(string check, PreType preType, IToken tok, string errorFormatString) {
+      confirmations.Add(() => {
+        if (!ConfirmConstraint(check, preType, null)) {
+          PreTypeResolver.ReportError(tok, errorFormatString, preType);
+        }
+      });
+    }
+
+    void AddConfirmation2(string check, PreType preType, Type toType, IToken tok, string errorFormatString) {
+      Contract.Requires(toType is NonProxyType);
+      var toPreType = (DPreType)PreTypeResolver.Type2PreType(toType);
+      confirmations.Add(() => {
+        if (!ConfirmConstraint(check, preType, toPreType)) {
+          PreTypeResolver.ReportError(tok, errorFormatString, preType);
+        }
+      });
+    }
+
+    void AddConfirmation(System.Action confirm) {
+      confirmations.Add(confirm);
+    }
+
+    void ConfirmTypeConstraints() {
+      foreach (var c in confirmations) {
+        c();
+      }
+    }
+
+    private bool ConfirmConstraint(string check, PreType preType, DPreType auxPreType) {
+      preType = preType.Normalize();
+      if (preType is PreTypeProxy) {
+        return false;
+      }
+
+      var pt = (DPreType)preType;
+      var ancestorPt = PreTypeResolver.NewTypeAncestor(pt);
+      var ancestorDecl = ancestorPt.Decl;
+      var familyDeclName = ancestorDecl.Name;
+      switch (check) {
+        case "InIntFamily":
```

Comment:
The usage of string literals here seems like a code-smell to me, but without he consuming code I'm not sure what alternative to suggest.

Reply:
I can change these string literals to an `enum`, but I prefer to do that in the large PR that that also contains the call sites.